### PR TITLE
Generate signature files for all input sources via compiler flag

### PR DESCRIPTION
**Pull request by baronfel**
This PR adds a compiler flag, config property, and implementation for a new compiler capability: the ability to generate all matching signature files for a lit of input sources. We already had the ability to generate a combined signature file, but this style of signature is not very useful from a design or tooling point of view, as it doesn't help with establishing module boundaries file-to-file within the project.

The new flag determines the name of the signature file from the name of the input source file, handling ml-style files by generating `.mli` and F# files by generating `.fsi`.

I've also annotated some of the options types with field names for better debugging.

### CODE REVIEW

**Comment by baronfel** [comment]
Alright, this is green now! Can I get a design/implementation review now? I think this would be incredibly useful (coupled with the xml-comment changes as well) for helping teams on-board legacy code with signature files, helping them take advantage of the IDE performance that has recently become available.

File: src/fsharp/fsc.fs [hunk 1]
```diff
@@ -324,23 +324,52 @@ module InterfaceFileWriter =
...
+                let filename = System.IO.Path.ChangeExtension(name.Range.FileName, extensionForFile name.Range.FileName)
+                printfn "writing impl file to %s" filename
+                use os = File.CreateText filename :> TextWriter
+                writeHeader filename os
+                writeToFile os impl
+
+        if tcConfig.printSignature
+        then writeAllToSameFile declaredImpls
+        else if tcConfig.printAllSignatureFiles
+        then writeToSeparateFiles declaredImpls
```
Comment (cartermp) [inline src/fsharp/fsc.fs:372]:
```suggestion
        if tcConfig.printSignature then
            writeAllToSameFile declaredImpls
        else if tcConfig.printAllSignatureFiles then
            writeToSeparateFiles declaredImpls
```

File: src/fsharp/fsc.fs [hunk 2]
```diff
@@ -324,23 +324,52 @@ module InterfaceFileWriter =
...
+
+            for impl in declaredImpls do
+                writeToFile os impl
+
+            if tcConfig.printSignatureFile <> "" then os.Dispose()
+
+        let extensionForFile (filePath: string) =
+            if (List.exists (Filename.checkSuffix filePath) mlCompatSuffixes)
+            then ".mli"
+            else ".fsi"
```
Comment (cartermp) [inline src/fsharp/fsc.fs:359]:
```suggestion
            if List.exists (Filename.checkSuffix filePath) mlCompatSuffixes then
                ".mli"
            else
                ".fsi"
```

File: src/fsharp/fsc.fs [hunk 3]
```diff
@@ -324,23 +324,52 @@ module InterfaceFileWriter =
...
+        let writeHeader filePath os =
+            if filePath <> "" && not (List.exists (Filename.checkSuffix filePath) FSharpLightSyntaxFileSuffixes) then
+                fprintfn os "#light"
+                fprintfn os ""
+
+        let writeAllToSameFile declaredImpls =
+            /// Use a UTF-8 Encoding with no Byte Order Mark
+            let os =
+                if tcConfig.printSignatureFile = "" then Console.Out
+                else (File.CreateText tcConfig.printSignatureFile :> TextWriter)
```
Comment (cartermp) [inline src/fsharp/fsc.fs:347]:
```suggestion
                if tcConfig.printSignatureFile = "" then
                    Console.Out
                else
                    File.CreateText tcConfig.printSignatureFile :> TextWriter
```

**Review by cartermp — APPROVED** [review]

**Review by TIHan — APPROVED** [review]
Since this adds a compiler flag, do we need to do anything else? I don't know if we do RFCs for compiler flags or not. Otherwise, the changes look good.

**Comment by cartermp** [comment]
No RFC needed for this IMO. Didn't need it for the deterministic flag.

**Pull request merged**